chainstate/tx-verifier/src/transaction_verifier/input_output_policy.rs

```diff
@@ -13,75 +13,773 @@
+            );
+            Ok(())
         }
     }
 }
 
-/// Indicates whether an output purpose can be used in a tx as an output
-fn is_valid_output_for_tx(output: &TxOutput) -> bool {
-    match output {
-        TxOutput::Transfer(_, _)
-        | TxOutput::LockThenTransfer(_, _, _)
-        | TxOutput::Burn(_)
-        | TxOutput::StakePool(_) => true,
-        TxOutput::ProduceBlockFromStake(_, _, _) => false,
-    }
-}
-
-fn check_inputs_can_be_used_in_tx(
+/// Not all `TxOutput` combinations can be used in a transaction.
+pub fn check_tx_inputs_outputs_purposes(
     tx: &Transaction,
     utxo_view: &impl utxo::UtxosView,
 ) -> Result<(), ConnectTransactionError> {
-    let can_be_spent = tx
+    let inputs = tx
         .inputs()
         .iter()
         .map(|input| {
             utxo_view
                 .utxo(input.outpoint())
                 .map_err(|_| utxo::Error::ViewRead)?
+                .map(|u| u.output().clone())
                 .ok_or(ConnectTransactionError::MissingOutputOrSpent)
         })
-        .collect::<Result<Vec<_>, _>>()?
-        .iter()
-        .all(|utxo| is_valid_input_for_tx(utxo.output()));
+        .collect::<Result<Vec<_>, _>>()?;
+
+    match inputs.as_slice() {
+        // no inputs
+        [] => return Err(ConnectTransactionError::MissingTxInputs),
+        // single input
+        [input] => match tx.outputs() {
+            // no outputs
+            [] => { /* do nothing, it's ok to burn outputs in this way */ }
+            // single output
+            [output] => {
+                ensure!(
+                    is_valid_one_to_one_combination(input, output),
+                    ConnectTransactionError::InvalidOutputTypeInTx
+                );
+            }
+            // multiple outputs
+            _ => {
+                let valid_input = are_inputs_valid_for_tx(std::slice::from_ref(input));
+                ensure!(valid_input, ConnectTransactionError::InvalidInputTypeInTx);
+                let valid_outputs = are_outputs_valid_for_tx(tx.outputs());
+                ensure!(
+                    valid_outputs,
+                    ConnectTransactionError::InvalidOutputTypeInTx
+                );
+            }
+        },
+        // multiple inputs
+        _ => {
+            let valid_inputs = are_inputs_valid_for_tx(inputs.as_slice());
+            ensure!(valid_inputs, ConnectTransactionError::InvalidInputTypeInTx);
+            let valid_outputs = are_outputs_valid_for_tx(tx.outputs());
+            ensure!(
+                valid_outputs,
+                ConnectTransactionError::InvalidOutputTypeInTx
+            );
+        }
+    };
 
-    utils::ensure!(
-        can_be_spent,
-        ConnectTransactionError::AttemptToSpendInvalidOutputType
-    );
     Ok(())
 }
 
-fn check_outputs_can_be_used_in_tx(tx: &Transaction) -> Result<(), ConnectTransactionError> {
-    let are_outputs_valid = tx.outputs().iter().all(is_valid_output_for_tx);
+#[rustfmt::skip]
```

Comment:
If you want this to keep the OR lines at the end, I think it's a rule we can add to rustfmt

Reply:
good point, added a rule that allows aligning `|` in match arms:
https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#match_arm_leading_pipes

Besides, I wanted the pair to fit in a single line, but since you removed OutputPurpose it's ok now.